src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java

```diff
@@ -759,21 +768,146 @@ public boolean isKnownDatasource( UUID studyId, String participantId, String dat
+                    logger.error("unable to delete participant {}: participant does not exist.", participantId);
+                }
+                participantsToRemove.add( participantEntityKeyId );
+            } else {
+                // if no participant Id: load all participants and add to set
+                chronicleDataApi
+                        .loadEntitySetData( participantsEntitySetId, FileType.json,
+                                userToken ).forEach(
+                        entity -> entity.get( ID_FQN ).forEach(
+                                personId -> participantsToRemove.add( UUID.fromString( personId.toString() ) )
+                        )
+                );
+            }
+
+            // Be super careful here that the mapping is one-to-one:
+            // don't delete neighbors that might have other neighbors/participants
+            Set<UUID> srcNeighborSetIds = ImmutableSet.of( deviceESID, dataESID, preProcessedESID, answersESID );
+            Set<UUID> dstNeighborSetIds = ImmutableSet.of( answersESID );
+            Map<UUID, Set<UUID>> toDeleteEntitySetIdEntityKeyId = new HashMap<>();
+
+            participantsToRemove.forEach(
+                    participantEntityKeyId -> {
+                        // Get neighbors
+                        Map<UUID, Map<UUID, SetMultimap<UUID, NeighborEntityIds>>> participantNeighbors = userSearchApi
+                                .executeFilteredEntityNeighborIdsSearch(
+                                        participantsEntitySetId,
+                                        new EntityNeighborsFilter(
+                                                Set.of( participantEntityKeyId ),
+                                                java.util.Optional.of( srcNeighborSetIds ),
+                                                java.util.Optional.of( dstNeighborSetIds ),
+                                                java.util.Optional.empty()
+                                        )
+                                );
+
+                        if ( participantNeighbors.size() == 0 ) {
+                            logger.debug( "Attempt to remove participant without data." );
+                        }
+
+                        // create a key for all entity sets
+                        Sets.union( srcNeighborSetIds, dstNeighborSetIds ).forEach( entitySetId -> {
+                            toDeleteEntitySetIdEntityKeyId.put( entitySetId, new HashSet<>() );
+                        } );
+
+                        // fill Map<entitySetId, Set<entityKeyId>>
+                        participantNeighbors
+                                .get( participantEntityKeyId )
+                                .forEach( ( edgeEntitySetId, edgeNeighbor ) -> {
+                                    edgeNeighbor.forEach( ( neighborEntitySetId, neighborEntityIds ) -> {
+                                        Set<UUID> set = toDeleteEntitySetIdEntityKeyId.get( neighborEntitySetId );
```

Comment:
What's the purpose of this set?

Reply:
The goal is to have a map:
Map<studyEntitySetId, Set<participantEntityKeyIds>>, eg.
```
id(chronicle_devices): 
- id(studyA_person1_device)
- id(studyA_person2_device)
- ...
id(chronicle_app_data): 
- id(studyA_person1_appdata1)
- id(studyA_person1_appdata2)
- id(studyA_person2_appdata1)
...
```

This way, we're collecting id's in the different entity sets that belong to *all* the participants to remove, to avoid having to loop through participants.

Reply:
But you're probably saying it shouldn't be stored as a set to add, which is now implemented as:
```
toDeleteEntitySetIdEntityKeyId.get( neighborEntitySetId )
    .add( neighborEntityIds.getNeighborEntityKeyId() );
```
:)

Reply:
ohh I see what you're doing -- I guess this does actually update the map. Reading through before I didn't see that this set was the one in the map and thought you were just adding to a set that would disappear once the loop exited.